File: bench/particle/particle.c
/* particle.c — fixed-timestep particle integrator (gaming physics step).
 * Bit-identical to particle.js: same XorShift32 seed, same semi-implicit Euler,
 * same FNV-1a-over-f64-bytes checksum. Built with -ffp-contract=off so the
 * `a + b*c` integration matches the scalar f64 reference (no FMA fusion). */
#include <stdint.h>
#include <stdio.h>
#include <string.h>
#include <time.h>

#define N (1 << 16)
#define STEPS 256
#define DT 0.015625
#define G (-9.8)
#define N_RUNS 21
#define N_WARMUP 5

static double px[N], py[N], vx[N], vy[N];

static uint32_t mix(uint32_t h, uint32_t x) { return (uint32_t)((uint64_t)(h ^ x) * 0x01000193u); }

static uint32_t checksum_f64(const double *out, int n) {
  uint32_t h = 0x811c9dc5u;
  int total_words = n * 2;
  for (int i = 0; i < total_words; i += 256) {
    int byte_off = i * (int)sizeof(uint32_t);
    uint32_t w;
    memcpy(&w, (const char *)out + byte_off, sizeof(w));
    h = mix(h, w);
  }
  return h;
}

static void seed_state(void) {
  uint32_t s = 0x1234abcd;
  for (int i = 0; i < N; i++) {
    s ^= s << 13; s ^= s >> 17; s ^= s << 5;
    px[i] = (double)s / 4294967296.0 * 2.0 - 1.0;
    s ^= s << 13; s ^= s >> 17; s ^= s << 5;
    py[i] = (double)s / 4294967296.0 * 2.0 - 1.0;
    s ^= s << 13; s ^= s >> 17; s ^= s << 5;
    vx[i] = (double)s / 4294967296.0 * 2.0 - 1.0;
    s ^= s << 13; s ^= s >> 17; s ^= s << 5;
    vy[i] = (double)s / 4294967296.0 * 2.0 - 1.0;
  }
}

static void step(void) {
  for (int i = 0; i < N; i++) {
    double nvy = vy[i] + G * DT;
    px[i] = px[i] + vx[i] * DT;
    py[i] = py[i] + nvy * DT;
    vy[i] = nvy;
  }
}

static void run(void) { for (int f = 0; f < STEPS; f++) step(); }

static double now_ms(void) {
  struct timespec ts;
  clock_gettime(CLOCK_MONOTONIC, &ts);
  return ts.tv_sec * 1000.0 + ts.tv_nsec / 1e6;
}

static uint64_t median_us(double *s, int n) {
  for (int i = 1; i < n; i++) {
    double v = s[i]; int j = i - 1;
    while (j >= 0 && s[j] > v) { s[j + 1] = s[j]; j--; }
    s[j + 1] = v;
  }
  return (uint64_t)(s[(n - 1) >> 1] * 1000.0);
}

int main(void) {
  for (int i = 0; i < N_WARMUP; i++) { seed_state(); run(); }
  double samples[N_RUNS];
  for (int i = 0; i < N_RUNS; i++) {
    seed_state();
    double t0 = now_ms();
    run();
    samples[i] = now_ms() - t0;
  }
  uint32_t cs = checksum_f64(py, N) ^ checksum_f64(px, N);
  printf("median_us=%llu checksum=%u samples=%d stages=%d runs=%d\n",
         (unsigned long long)median_us(samples, N_RUNS), cs, N * STEPS, STEPS, N_RUNS);
  return 0;
}

File: bench/particle/particle.rs
// particle.rs — fixed-timestep particle integrator. Bit-identical to particle.js.
use std::time::Instant;

const N: usize = 1 << 16;
const STEPS: usize = 256;
const DT: f64 = 0.015625;
const G: f64 = -9.8;
const N_RUNS: usize = 21;
const N_WARMUP: usize = 5;

fn mix(h: u32, x: u32) -> u32 {
    (h ^ x).wrapping_mul(0x0100_0193)
}

fn checksum(out: &[f64]) -> u32 {
    let mut h = 0x811c_9dc5u32;
    for i in (0..out.len() * 2).step_by(256) {
        let bytes = out[i / 2].to_le_bytes();
        let off = (i & 1) * 4;
        let w = u32::from_le_bytes([bytes[off], bytes[off + 1], bytes[off + 2], bytes[off + 3]]);
        h = mix(h, w);
    }
    h
}

fn seed_state(px: &mut [f64], py: &mut [f64], vx: &mut [f64], vy: &mut [f64]) {
    let mut s = 0x1234abcdu32;
    let mut r = || {
        s ^= s << 13;
        s ^= s >> 17;
        s ^= s << 5;
        (s as f64) / 4294967296.0 * 2.0 - 1.0
    };
    for i in 0..N {
        px[i] = r();
        py[i] = r();
        vx[i] = r();
        vy[i] = r();
    }
}

fn step(px: &mut [f64], py: &mut [f64], vx: &mut [f64], vy: &mut [f64]) {
    for i in 0..N {
        let nvy = vy[i] + G * DT;
        px[i] = px[i] + vx[i] * DT;
        py[i] = py[i] + nvy * DT;
        vy[i] = nvy;
    }
}

fn run(px: &mut [f64], py: &mut [f64], vx: &mut [f64], vy: &mut [f64]) {
    for _ in 0..STEPS {
        step(px, py, vx, vy);
    }
}

fn median_us(samples: &mut [f64]) -> u64 {
    for i in 1..samples.len() {
        let v = samples[i];
        let mut j = i;
        while j > 0 && samples[j - 1] > v {
            samples[j] = samples[j - 1];
            j -= 1;
        }
        samples[j] = v;
    }
    (samples[(samples.len() - 1) >> 1] * 1000.0) as u64
}

fn main() {
    let mut px = vec![0.0f64; N];
    let mut py = vec![0.0f64; N];
    let mut vx = vec![0.0f64; N];
    let mut vy = vec![0.0f64; N];

    for _ in 0..N_WARMUP {
        seed_state(&mut px, &mut py, &mut vx, &mut vy);
        run(&mut px, &mut py, &mut vx, &mut vy);
    }

    let mut samples = [0.0f64; N_RUNS];
    for s in &mut samples {
        seed_state(&mut px, &mut py, &mut vx, &mut vy);
        let t0 = Instant::now();
        run(&mut px, &mut py, &mut vx, &mut vy);
        *s = t0.elapsed().as_secs_f64() * 1000.0;
    }

    let cs = checksum(&py) ^ checksum(&px);
    println!(
        "median_us={} checksum={} samples={} stages={} runs={}",
        median_us(&mut samples),
        cs,
        N * STEPS,
        STEPS,
        N_RUNS
    );
}
